src/tqec/compile/block.py

```diff
@@ -0,0 +1,145 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+
+import cirq
+import cirq.circuits
+
+from tqec.circuit.circuit import generate_circuit
+from tqec.exceptions import TQECException
+from tqec.plaquette.library.empty import empty_square_plaquette
+from tqec.plaquette.plaquette import Plaquette, Plaquettes, RepeatedPlaquettes
+from tqec.position import Position2D
+from tqec.templates.base import RectangularTemplate
+from tqec.templates.scale import LinearFunction
+from tqec.templates.tiled import TiledTemplate
+
+
+@dataclass
+class CompiledBlock:
+    """Represents a specific implementation of a cube in a `BlockGraph`.
+
+    Attributes:
+        template: the template that defines the cube implementation.
+        layers: a list of `Plaquettes` that represent different functional layers of the
+            cube. When aligning two `CompiledBlock`s, the layers are aligned in order.
+            Typically, there are three layers in most cube implementations:
+            Initialization, Repetitions, and Measurement.
+    """
+
+    template: RectangularTemplate
+    layers: list[Plaquettes]
+
+    @property
+    def num_layers(self) -> int:
+        return len(self.layers)
+
+    def with_updated_layer(
+        self,
+        plaquettes_to_update: dict[int, Plaquette],
+        layer_to_update: int,
+    ) -> CompiledBlock:
+        """Returns a new `CompiledBlock` with the specified layer updated.
+
+        Args:
+            plaquettes_to_update: a dictionary of plaquettes to update in the layer.
+            layer_to_update: the index of the layer to update.
+        """
+        new_plaquette_layers = []
+        for i, plaquettes in enumerate(self.layers):
+            if i == layer_to_update:
+                new_plaquette_layers.append(
+                    plaquettes.with_updated_plaquettes(plaquettes_to_update)
+                )
+            else:
+                new_plaquette_layers.append(plaquettes)
+        return CompiledBlock(self.template, new_plaquette_layers)
+
+
+class TiledBlocks:
+    """Represents a collection of `CompiledBlock`s tiled in a 2D grid."""
+
+    def __init__(self, blocks_by_position: dict[Position2D, CompiledBlock]):
+        templates_by_position: dict[Position2D, RectangularTemplate] = {}
+        for pos, block in blocks_by_position.items():
+            templates_by_position[pos] = block.template
+        self._tiled_template = TiledTemplate(template_by_position=templates_by_position)
+        self._tiled_layers = self._merge_layers(
+            {pos: block.layers for pos, block in blocks_by_position.items()}
+        )
+
+    @property
+    def block_size(self) -> int:
+        """Return the uniform block size of the tiled blocks.
+
+        In this implementation, all the block itself should be square
+        shape and have the same width and height as the block size. And
+        all the blocks should have the same size.
```

Comment:
That should probably be asserted somewhere, with an exception raised if that is not the case.